OrbitLinuxTracing/GpuTracepointEventProcessor.cpp

```diff
@@ -0,0 +1,179 @@
+#include "GpuTracepointEventProcessor.h"
+
+#include <string>
+#include <tuple>
+#include <vector>
+
+namespace LinuxTracing {
+
+namespace {
+
+// Format is based on the content of the event's format file:
+// /sys/kernel/debug/tracing/events/<category>/<name>/format
+struct __attribute__((__packed__)) perf_event_amdgpu_cs_ioctl {
+  uint16_t common_type;
+  uint8_t common_flags;
+  uint8_t common_preempt_count;
+  int32_t common_pid;
+  uint64_t sched_job_id;
+  int32_t timeline;
+  uint32_t context;
+  uint32_t seqno;
+  uint64_t dma_fence;  // This is an address.
+  uint64_t ring_name;  // This is an address.
+  uint32_t num_ibs;
+};
+
+struct __attribute__((__packed__)) perf_event_amdgpu_sched_run_job {
+  uint16_t common_type;
+  uint8_t common_flags;
+  uint8_t common_preempt_count;
+  int32_t common_pid;
+  uint64_t sched_job_id;
+  int32_t timeline;
+  uint32_t context;
+  uint32_t seqno;
+  uint64_t ring_name;  // This is an address.
+  uint32_t num_ibs;
+};
+
+struct __attribute__((__packed__)) perf_event_dma_fence_signaled {
+  uint16_t common_type;
+  uint8_t common_flags;
+  uint8_t common_preempt_count;
+  int32_t common_pid;
+  int32_t driver;
+  int32_t timeline;
+  uint32_t context;
+  uint32_t seqno;
+};
+
+}  // namespace
+
+int GpuTracepointEventProcessor::ComputeDepthForEvent(
+    const std::string& timeline,
+    uint64_t start_timestamp, uint64_t end_timestamp) {
+  if (timeline_to_latest_timestamp_per_depth_.count(timeline) == 0) {
+    std::vector<uint64_t> vec;
+    timeline_to_latest_timestamp_per_depth_.emplace(timeline, vec);
+  }
+  auto it = timeline_to_latest_timestamp_per_depth_.find(timeline);
+  std::vector<uint64_t>& vec = it->second;
+
+  for (int d = 0; d < vec.size(); ++d) {
+    // We add a small slack on each row of the GPU track timeline to
+    // make sure events don't get too crowded.
+    constexpr uint64_t slack_ns = 5 * 1000000;
+    if (start_timestamp >= (vec[d] + slack_ns)) {
+      vec[d] = end_timestamp;
+      return d;
+    }
+  }
+  vec.push_back(end_timestamp);
+  return static_cast<int>(vec.size());
+}
+
+void GpuTracepointEventProcessor::CreateGpuExecutionEventIfComplete(
```

Comment:
Is it possible to get these 3 events out of order?  Is that why we don't "CreateGpuExecutionEventIfComplete" only on the last event?

Reply:
Yes, this is possible and I have observed this. I have added a comment to clarify.